Add argument parsing, `add_console_command` and other features

tqwewe opened the pull request:
This PR is a breaking change and adds a lot of things including:

 - Argument parsing
 - Registering commands with `app.add_console_command`
 - Derive macro for defining console commands with structs
 - Console command help information and usage
 - `ConsoleCommand` system parameter
 - Built in `help`, `clear` and `exit` console commands
 - Command history with up and down arrows
 
There's a lot to go through, the best place to start would be to clone the fork and run the examples.

Basic example of defining and handling a console command:

```rust
/// Prints given arguments to the console
#[derive(ConsoleCommand)]
#[console_command(name = "log")]
struct LogCommand {
    /// Message to print
    msg: String,
    /// Number of times to print message
    num: Option<i64>,
}

fn log_command(mut log: ConsoleCommand<LogCommand>) {
    if let Some(LogCommand { msg, num }) = log.take() {
        let repeat_count = num.unwrap_or(1);

        for _ in 0..repeat_count {
            reply!(log, "{msg}");
        }

        log.ok();
    }
}
```
```rust
app.add_console_command::<LogCommand, _, _>(log_command);
```

This code snippet will register the command when the user enters `help` and `help log` (using doc comments), and will only execute if the arguments are parsed successfully into the correct types.

<img width="410" alt="Screen Shot 2022-02-23 at 6 34 46 pm" src="https://user-images.githubusercontent.com/16362377/155311826-1ab2c34b-a1e3-4397-aefd-7ea64343b2d1.png">

alice-i-cecile commented on src/lib.rs:
```diff
@@ -1,319 +1,40 @@
+            .add_console_command::<ExitCommand, _, _>(exit_command)
+            .add_console_command::<HelpCommand, _, _>(help_command)
+            // if there's other egui code we need to make sure they don't run at the same time
+            .add_system(console_system.exclusive_system())
```
I'm not sure this is needed; don't you take a `ResMut` lock on the `egui` resource?

tqwewe replied:
Seems like this isn't needed actually. I left this as-is from before my changes, but it seems like I might've removed the need for it somewhere along the way.

Will remove it in my next commit

alice-i-cecile commented on src/value.rs:
```diff
@@ -0,0 +1,179 @@
+pub enum ValueType {
+    /// String arg type
+    String,
+    /// Int arg type
```
What type of integer are these stored as?

tqwewe replied:
Int is stored as `i64`, and Float is stored as `f64`
https://github.com/RichoDemus/bevy-console/pull/4/files/b2d9dfd423fbf978c03794f00f4d0ac3f954f977#diff-df3933aef2020e10367701a3160b4c9efc31e32d7202019c8a17de0f267da116R20-R21

You can define a command with
 - `u8` up to `u128` (including `usize`)
 - `i8` up to `i128` (including `isize`)
 - `f32` and `f641
 
But these values will be casted using `as`. Maybe I should remove `128` bit types, and any types that could result in a potentially lossy cast.
https://github.com/RichoDemus/bevy-console/pull/4/files/b2d9dfd423fbf978c03794f00f4d0ac3f954f977#diff-be1c878c753064b7eaddf4a10ba4215a93dd1817d31e6788691cdb29752fd612R117

alice-i-cecile replied:
Good; I think these are sizable defaults. I agree with removing 128 bit types: they're too niche and will fail surprisingly. This comment should be documented on those fields :)

tqwewe replied:
Honestly, i64 can only be safely casted as i128. I think I should remove the implementations and only implement `i64` and `f64` to avoid any unsafe casting.
If the user wants to cast, they can put i64/f64 in their struct and cast manually I think.

So this means the user cannot put `u8` and such in their command arguments

alice-i-cecile commented:
First pass on code quality looks good! I'm generally impressed with the features and ergonomics of the design.

RichoDemus commented:
Thanks for the PR and for all the comments, I'll try to take a look this week!

tqwewe commented:
Made some updates including:
- Command history with up and down arrows
- Fixed terminal scroll width not expanding to fill window
- Fixed scroll glitching when you scroll up
- And some other small updates

tqwewe commented:
I also tried tackling the issue https://github.com/RichoDemus/bevy-console/issues/2 of blocking input when the console is open by clearing the input events, but doing so made the console terminal break too not allowing you to press enter.

This issue is a difficult one to solve which I'm not sure is really possible.

RichoDemus commented:
This looks very good, thanks a lot for your effort. I'm very positive to merging this, I just have a few things
1. There are some pieces of code that are commented out, are they old things that should be removed?
2. So right now the command description is generated from the comment, I think that's a bit to "magical" and that it would be a lot more clear if it was a macro parameter just like `name` is now, what do you think?
I also think that maybe it could be desirable to be able to actually write a comment for a command without that being exposed to the end-user

tqwewe commented:
Thanks for the feedback!

> 1. There are some pieces of code that are commented out, are they old things that should be removed?

I had a look through but couldn't really see any old code commented out. Perhaps you just saw my doc comment examples and mistook it for old code I commented out?

The only commented code I could find, I've removed in this commit: https://github.com/tqwewe/bevy-console/commit/5458546e4a9d617579fa82150a704d454f9c6004

> 2. So right now the command description is generated from the comment, I think that's a bit to "magical" and that it would be a lot more clear if it was a macro parameter just like name is now, what do you think?

I agree that it's a little magical, but I think that's just part of the power of Rust macros. I don't really feel that it's overly magical to confuse the user though, it's just a feature of macros imo.

If it was an attribute, then what if you also wanted doc comments? It might be a little frustrating to have to write your info twice.
Eg:
```rust
/// Prints given arguments to the console
#[derive(ConsoleCommand)]
#[console_command(name = "log")]
#[console_command(help = "Prints given arguments to the console")]
pub struct LogCommand {
    /// Message to print
    #[console_command(help = "Message to print")]
    pub msg: String,
    /// Number of times to print message
    #[console_command(help = "Number of times to print message")]
    pub num: Option<i64>,
}
```
I feel that this just adds too much boilerplate when the help would almost always be the same as the doc comment right? And you also have the annoying thing of the docs going out of sync with the help attribute, whereas having it all as a doc comment kind of kills two birds with one stone without many disadvantages.

> I also think that maybe it could be desirable to be able to actually write a comment for a command without that being exposed to the end-user
If that's true then it is indeed a down side. I'm curious to hear if @alice-i-cecile has any thoughts on this?

TAINCER commented:
I looked through the new commands and found you made inconsistent use of `help.reply()` and the `reply!()` macro in the [help](https://github.com/tqwewe/bevy-console/blob/5458546e4a9d617579fa82150a704d454f9c6004/src/commands/help.rs) command. Is there a reason for this? Really looking forward to when this is merged.

RichoDemus commented on bevy_console_parser/src/value.rs:
```diff
@@ -44,17 +44,6 @@ impl From<ValueRaw<'_>> for Value {
-//         }
-//     }
-// }
-
```
I think this was the commented code I saw :)

tqwewe replied:
Ah okay, I've removed that in a recent commit

RichoDemus commented:
Oh yeah, and `cargo test` fails
I should add some simple github action ^^

tqwewe commented:
I've just fixed the doc tests, so it should all be passing now
